Add ROCm utilities and tests, and CUDA fallbacks

fwyzard opened the pull request:
#### PR description:

Add basic ROCm functions:
  - `hipCheck()` macro to check if HIP API functions are successful, or raise an exception;
  - `cms::rocmtest::requireDevices()` function to check if the system has any ROCm GPUs, or exit successfully.

Add tests for these functions and for the corresponding CUDA functions.

Add basic ROCm utilities:
  - `rocmIsEnabled` exits with success if a ROCm device is available ad configured correctly
  - `rocmComputeCapabilities` lists all ROCm devices and their GCN architecture:
    ```bash
    $ rocmComputeCapabilities 
       0      gfx900    AMD Radeon Pro WX 9100
    ```

#### PR validation:

The new unit tests pass on machines with and without GPUs.

fwyzard committed:
Add a test for the cudaCheck() macro

fwyzard committed:
Implement the hipCheck() macro

fwyzard committed:
Implement requireDevices() for ROCm devices

fwyzard committed:
Add a test for requireDevices() for CUDA devices

fwyzard commented:
please test with cms-sw/cmsdist#8266

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33893

ERROR: Build errors found during clang-tidy run.
```
#error("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
 ^
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/el8_amd64_gcc11/external/rocm/5.0.2-95c215630c939706b0552e3eee38861c/include/hip/hip_runtime_api.h:5501:2: error: ("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__"); [clang-diagnostic-error]
#error("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
 ^
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/el8_amd64_gcc11/external/rocm/5.0.2-95c215630c939706b0552e3eee38861c/include/hip/hip_runtime_api.h:5526:61: error: use of undeclared identifier 'hipHostMallocDefault' [clang-diagnostic-error]
                                       unsigned int flags = hipHostMallocDefault) {
                                                            ^
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/el8_amd64_gcc11/external/rocm/5.0.2-95c215630c939706b0552e3eee38861c/include/hip/hip_runtime_api.h:5532:61: error: use of undeclared identifier 'hipMemAttachGlobal' [clang-diagnostic-error]
                                       unsigned int flags = hipMemAttachGlobal) {
                                                            ^
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/el8_amd64_gcc11/external/rocm/5.0.2-95c215630c939706b0552e3eee38861c/include/hip/hip_vector_types.h:38:2: error: ("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__"); [clang-diagnostic-error]
#error("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
 ^
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02769/el8_amd64_gcc11/external/rocm/5.0.2-95c215630c939706b0552e3eee38861c/include/hip/library_types.h:48:2: error: ("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__"); [clang-diagnostic-error]
#error("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
 ^
HeterogeneousCore/ROCmUtilities/src/requireDevices.cc:12:19: error: use of undeclared identifier 'hipGetDeviceCount' [clang-diagnostic-error]
    auto status = hipGetDeviceCount(&devices);
                  ^
Suppressed 213 warnings (212 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

fwyzard commented:
@makortel these are the basic HIP/ROCm utilities I would add, to ease the implementation of the ROCm service and build rules.

However, we should settle on the "HIP" vs "ROCm" naming approach...

As far as I understand, "ROCm" indicates the backend and the kind of hardware, while "HIP" indicates the API.

Which makes sense: for example, one could program ROCm devices using other APIs, like OpenCL or SYCL, or use HIP to program other devices, like NVIDIA GPUs.

On the other hand, at the moment we do not plan to use HIP on other hardware than ROCm -- though _HIP-over-Level 0_ is an interesting idea to avoid SYCL...

So far I've used
  - HIP wherever the API is the most relevant aspect, _e.g._ `hipCheck(...)`;
  - ROCm wherever the architecture is the most relevant aspect, _e.g._ the `rocm` SCRAM tool and the `cms::rocmtest::requireDevices()` function.

However, I guess we could also just use "HIP" everywhere (except maybe the tool).


What do you think ?

fwyzard commented:
> -code-checks

I guess the code checks do not use the flags from cms-sw/cmsdist#8266 ?

smuzaffar commented:
code-checks with cms.week1.PR_92dfa22c/53.0-3fa13e8f691137f6726f64f123800e88

lets run the code-checks using https://github.com/cms-sw/cmsdist/pull/8266 tool-conf

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33895

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAUtilities/test/BuildFile.xml  modified in PR(s): #37952, #35713

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33895/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33895/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented:
code-checks with cms.week1.PR_92dfa22c/53.0-3fa13e8f691137f6726f64f123800e88

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33898

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAUtilities/test/BuildFile.xml  modified in PR(s): #37952, #35713

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

- HeterogeneousCore/CUDAUtilities (**heterogeneous**)
- HeterogeneousCore/ROCmUtilities (****)


The following packages do not have a category, yet:

HeterogeneousCore/ROCmUtilities
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@cmsbuild, @makortel, @fwyzard can you please review it and eventually sign? Thanks.
@makortel, @missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fwyzard committed:
Implement rocmIsEnabled and rocmComputeCapabilities

makortel commented:
> What do you think ?

@fwyzard I pretty much agree with what you wrote. From a technical standpoint keeping the HIP and ROCm concepts separate (as you did here) would help to keep things clear. Let me nevertheless think out loud a bit (for future reference if nothing more).

On the other hand, on a cursory look all code currently in `HeterogeneousCore/ROCmUtilities` uses the HIP API, and I'd naively expect that code itself to work on any HIP's "backend" without changes. So from that perspective the package could be named as `HeterogeneousCore/HIPUtilities`.

Then a counter-OTOH, the package explicitly uses the ROCm backend of HIP, and therefore one can't really the "HIP utilities" without ROCm. Given, as you wrote, that we are not planning to use HIP directly (but as one Alpaka backend). maybe the best point of view for our use case would be along "the package is about using ROCm 'directly', and that 'happens' to be done via HIP". I think this point of view corresponds to what you did here.

Since we expect the direct use of HIP/ROCm would be limited, I'd expect changing nomenclature in the future to be fairly straightforward, and therefore I don't see a big risk in making a suboptimal decision now.

I suppose in any (grid) resource interfaces we would anyway use the "ROCm" name because those would be strictly about using AMD GPUs.

fwyzard commented:
OK, then I'll leave the names as they are, at least for the time being.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33908

ERROR: Build errors found during clang-tidy run.
```
HeterogeneousCore/ROCmServices/bin/isRocmDeviceSupported.hip.cc:29:17: error: expected expression [clang-diagnostic-error]
  setSupported<<<1, 1>>>(supported_d);
                ^
HeterogeneousCore/ROCmServices/bin/isRocmDeviceSupported.hip.cc:29:24: error: expected expression [clang-diagnostic-error]
  setSupported<<<1, 1>>>(supported_d);
                       ^
Suppressed 97 warnings (96 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

makortel commented on HeterogeneousCore/ROCmServices/bin/BuildFile.xml:
```diff
@@ -0,0 +1,9 @@
+    <use name="rocm"/>
+  </bin>
+
+  <bin name="rocmIsEnabled" file="rocmIsEnabled.cpp isRocmDeviceSupported.hip.cc">
```
I had long thought we would (eventually) always build `cudaIsEnabled`, `rocmIsEnabled` etc, and in case the corresponding platform is not available on a given `SCRAM_ARCH`, the program would just return `EXIT_FAILURE`.

On the other hand, for the use in python configuration, the failure to run such a program from `$PATH` would be automatically interpreted as "platform is unavailable", so maybe this current approach of not building the `*IsEnabled` program when the platform is not available is ok.

Thoughts?

fwyzard replied:
I'm fine with always building something, that would just return "false".
Does SCRAM support something like `<ifnottool>` ?

smuzaffar replied:
> Does SCRAM support something like <ifnottool> ?

yes SCRAM does support it but you need to use `!` e.g `<iftool name="!rocm"/>`

fwyzard replied:
mhm, I tried this, but I couldn't get it to work :-(
do we have any examples ?

smuzaffar replied:
ah my bad, `!` was supported for all the other filters (e.g ifrelease, ifarch etc) but not for iftool. https://github.com/cms-sw/cmsdist/pull/8272 should fix this

fwyzard replied:
thanks !

fwyzard replied:
@smuzaffar ok, so with this can have two different source files, one that uses ROCm and one that does not:
```xml
<iftool name="rocm">
  <bin name="rocmIsEnabled" file="rocmIsEnabled.cpp isRocmDeviceSupported.hip.cc">
    <use name="rocm"/>
  </bin>
</iftool>
<iftool name="!rocm">
  <bin name="rocmIsEnabled" file="rocmIsEnabled_fallback.cpp">
  </bin>
</iftool>
```
Is that OK ?

smuzaffar replied:
you can use `<else/>` tag e.g. the following should work with current SCRAM

```
<iftool name="rocm">
  <bin name="rocmIsEnabled" file="rocmIsEnabled.cpp isRocmDeviceSupported.hip.cc">
    <use name="rocm"/>
  </bin>
<else/>
  <bin name="rocmIsEnabled" file="rocmIsEnabled_fallback.cpp">
  </bin>
</iftool>
```

smuzaffar commented:
code-checks with cms.week1.PR_6b9a106c/53.0-39caaec466106b0f55d488374202f19f

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33913

ERROR: Build errors found during clang-tidy run.
```
HeterogeneousCore/ROCmServices/bin/isRocmDeviceSupported.hip.cc:29:17: error: expected expression [clang-diagnostic-error]
  setSupported<<<1, 1>>>(supported_d);
                ^
HeterogeneousCore/ROCmServices/bin/isRocmDeviceSupported.hip.cc:29:24: error: expected expression [clang-diagnostic-error]
  setSupported<<<1, 1>>>(supported_d);
                       ^
Suppressed 97 warnings (96 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

smuzaffar commented:
@fwyzard , is clang-tidy suppose to work for `.hip.cc` files?

fwyzard commented:
I have not tried.

I would _guess_ it should work, but it may require extra flags to indicate HIP code.

fwyzard commented:
code-checks with cms.week1.PR_6b9a106c/53.0-773f204b2827f2767588a4268e6f0d6a

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33915

ERROR: Build errors found during clang-tidy run.
```
HeterogeneousCore/ROCmServices/bin/isRocmDeviceSupported.hip.cc:29:17: error: expected expression [clang-diagnostic-error]
  setSupported<<<1, 1>>>(supported_d);
                ^
HeterogeneousCore/ROCmServices/bin/isRocmDeviceSupported.hip.cc:29:24: error: expected expression [clang-diagnostic-error]
  setSupported<<<1, 1>>>(supported_d);
                       ^
Suppressed 97 warnings (96 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33918

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAUtilities/test/BuildFile.xml  modified in PR(s): #37952, #35713

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33918/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33918/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

smuzaffar commented:
@fwyzard , I have merged https://github.com/cms-sw/cmsdist/pull/8268 which should have working build rules for rocm. I noticed that we are still missing tests to build cuda and rocm based shared libs (and extra archive libs), can you please add some codes to build shared libs for cuda/rocm and some test/bin/plugins to depend on those?

fwyzard commented:
Yes, the idea is to implement the equivalent of #40605 for ROCm.

fwyzard commented:
please test

smuzaffar commented:
@fwyzard , you need to apply the code-format changes first

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33933

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAUtilities/test/BuildFile.xml  modified in PR(s): #37952, #35713

cmsbuild commented:
Pull request #40619 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard committed:
Implement fallback for ROCm utilities

Implement fallback binaries for rocmIsEnabled and rocmComputeCapabilities
for platforms that do not support ROCm.

fwyzard committed:
Move CUDA utilities and implement fallbacks

Move the CUDA command line utilities cudaComputeCapabilities and
cudaIsEnabled from HeterogeneousCore/CUDAServices to CUDAUtilities.

Implement fallback binaries for cudaComputeCapabilities and cudaIsEnabled
for platforms that do not support CUDA.

fwyzard commented:
please test with cms-sw/cms-bot#1928,cms-sw/cmsdist#8274

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33940

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAUtilities/test/BuildFile.xml  modified in PR(s): #37952, #35713

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33940/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33940/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented:
please test with cms-sw/cms-bot#1928,cms-sw/cmsdist#8274

fwyzard commented:
+heterogeneous

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40619/33942

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDAUtilities/test/BuildFile.xml  modified in PR(s): #37952, #35713

cmsbuild commented:
Pull request #40619 was updated.  can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1f2297/30252/summary.html
**COMMIT**: bbcce061553c7daea332040ff59a2079efb131c1
**CMSSW**: CMSSW_13_0_X_2023-01-30-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40619/30252/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @bsunanda cms-sw/cmssw#40632
 - @bsunanda cms-sw/cmssw#40633
 - @missirol cms-sw/cmssw#40635
 - @mmusich cms-sw/cmssw#40608
 - @cms-sw cms-sw/cmsdist#8272

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1f2297/30252/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1f2297/30252/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 3 lines to the logs
 - Reco comparison results: 12 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555495
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555467
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-30-1100+1f2297/55228/triggerResults): no differences found

fwyzard commented:
please test with cms-sw/cmsdist#8273

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

fwyzard commented:
please test with cms-sw/cmsdist#8273

fwyzard commented:
please abort

fwyzard commented:
please test with cms-sw/cmsdist#8273

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1f2297/30279/summary.html
**COMMIT**: bbcce061553c7daea332040ff59a2079efb131c1
**CMSSW**: CMSSW_13_0_X_2023-01-31-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40619/30279/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @bsunanda cms-sw/cmssw#40638
 - @cms-sw cms-sw/cmsdist#8275

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1f2297/30279/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1f2297/30279/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 6 lines to the logs
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555495
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555464
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-31-1100+1f2297/55266/triggerResults): no differences found

perrotta commented:
+1
